Add `delete_pr_comment`

crazytonyli opened the pull request:
A new script to delete PR comments that contain a given content.

### Test instructions

Go to this PR's buildkite job page, select a step, and go to the "Environment" tab. Click the "Show export prefix" button and copy the environment variables. Open a terminal on your computer and paste the copied env vars. Invoking this script from your shell session should delete comments (or not) based on the `--containing` argument.

---

- [x] I have considered if this change warrants release notes and have added them to the appropriate section in the `CHANGELOG.md` if necessary.

crazytonyli committed:
Add `delete_pr_comment`

crazytonyli committed:
Add changelog

crazytonyli commented:
test comment

crazytonyli commented:
test comment again

crazytonyli commented:
comment

mokagio commented on bin/delete_pr_comment:
```diff
@@ -0,0 +1,48 @@
+#!/bin/bash -eu
+
+# This script is used to delete comments on a pull request, and must be run in a buldkite PR step.
```
```suggestion
# This script is used to delete comments on a pull request, and must be run in a Buildkite PR step.
```

mokagio commented:
Test comment 1

mokagio commented:
Test comment 2

mokagio commented:
Comment test 2

mokagio commented:
Comment test 3

mokagio commented:
Comment test 3

mokagio commented on bin/delete_pr_comment:
```diff
@@ -0,0 +1,48 @@
+GITHUB_REPO=$(basename "$GITHUB_URL")
+
+COMMENT_IDS="$(curl -s \
+    -H "Authorization: token ${GITHUB_TOKEN}" \
```
Just curious. Is there any advantage in interpolating with `${GITHUB_TOKEN}` over `$GITHUB_TOKEN`?

crazytonyli replied:
I guess `${}` is less prone to error, since it clearly states what the variable is. Like `${GITHUB_TOKEN}_VALUE` vs `$GITHUB_TOKEN_VALUE`. I think I copied and pasted a few code here and there, and simply forgot to make them consistent. 🥲

mokagio commented on bin/delete_pr_comment:
```diff
@@ -0,0 +1,48 @@
+}
+
+require_option '--containing' "$COMMENT_CONTAINING"
+require_option '--github-token' "$GITHUB_TOKEN"
```
I tried to break this by calling without passing the option and it worked anyway, my guess is because I have `$GITHUB_TOKEN` as an export in my environment.

<img width="1280" alt="Screenshot 2023-05-03 at 1 19 57 pm" src="https://user-images.githubusercontent.com/1218433/235826798-4425b365-78aa-4f8c-b8da-472832c38275.png">

I don't think this is a blocker in the context of how we can expect this script to be used.

crazytonyli replied:
Yeah, this is an intended, however undocumented, behaviour.

When invoking this script on CI, we have a few env vars defined by convention, GITHUB_TOKEN being one of them. They are used by many other tools too (`gh`, octokit etc.), it sometimes can be quite handy to just use the env var if it's not passed via CLI. 😄

mokagio commented on bin/delete_pr_comment:
```diff
@@ -0,0 +1,48 @@
+    fi
+}
+
+require_option '--containing' "$COMMENT_CONTAINING"
```
Because of the `u` option in the shebang, this call will fail before `require_option` gets a chance to run if `$COMMENT_CONTAINING` is empty:

```
./bin/delete_pr_comment: line 29: COMMENT_CONTAINING: unbound variable
```

The result is the same (the script fails) but the UX is not as clear.

crazytonyli replied:
Yeah I agree. Considering we have so many bash script, I thought I'll add this simply argument parsing function to give the scripts a better CLI, with `cli --foo value --bar true` instead of `cli value true`. It does its job, but the error message may not be intuitive.

mokagio commented on bin/delete_pr_comment:
```diff
@@ -0,0 +1,48 @@
+
+# Check dependencies and print their versions for diagnosis purposes
+jq --version
+curl --version
```
How do you feel about adding an empty line here so that the version diagnostics info are separated from the rest of the script output?

```suggestion
curl --version
# Print line to separate versions info from the rest of the script output
echo ''
```

crazytonyli replied:
Updated in https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/52/commits/5ad7153d746b7a0a20287cafc284707e2784bf89

mokagio reviewed (approved):
I tried it on some tests comments in this PR. Works as advertised 👍 

I'm going to approve to unblock, because nothing I said here is really a blocker.

However, there is one last thing I'd like to ask. I am a bit worried about unintended consequences of selecting any comment matching the given `containing` value. I'm probably overthinking this, because the usage of this script will be quite limited and I assume we'll also introduce our system to use unique identifiers or something (e.g. Danger leaves `<!-- comment id: 1234 -->`-style comments that do not render but are still readable to API calls). Still, I can see someone accidentally passing the wrong string to this command and deleting other people's comments 🤔

crazytonyli committed:
Refine the jq and curl dependency check output

crazytonyli committed:
Update a code comment

Co-authored-by: Gio Lodi <gio.lodi@automattic.com>

crazytonyli committed:
Restrict the delete_pr_comment to query by invisble block

crazytonyli commented:
@mokagio Regarding your concern on deleting other comment by accident. I definitely agree with you. That's on us to make sure the query criteria is unique enough, i.e. using an invisible code block like you mentioned. I extracted this script from an automation tool in one of our private repo. That tool posts and deletes comments that has certain invisible block content.

But I have added a restriction in https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/52/commits/0cdfa64dc528ea9c486b6f1f1a02e1338b9af15b, to only allow query `<!-- xxx -->`, which should reduce the chances of deleting wrong comments.

crazytonyli committed:
Update changelog

crazytonyli commented:
@AliSoftware One side effect with this approach is, the previous comment, which potentially contains outdated and inaccurate information, would only be deleted at the time a new comment is updated or posted. Continuing with your example, let's say the `terraform plan` step failed, what you'd see on PR is a failed "terraform-plan" CI check and a comment saying "here is the terraform plan" with an outdated plan.

In the PR where this `delete_pr_comment` was originated from, I initially took the "updating the existing comment" approach you mentioned, but ended up going with deleting the comment first before creating a new one, because it can produce the most accurate information.

But I'm happy to remove this `delete_pr_comment` from this plugin though, deleting a comment probably isn't a common thing we do. It should be fine just keep the script in the repo that needs it.

AliSoftware commented:
That's a valid point.

And to be fair that's a limitation that we already have with the comments we do on PRs of app repos to announce where to download Prototype Build for the PR: if the first CI build succeeded in building the Prototype Build, but q subsequent commit and need CI build failed to build the Prototype Build on the new commit, the old comment pointing to the old Prototype Build would not be dismissed while in practice it would not represent the reality of "the Prototype Build containing every changes that appears on this PR" (but instead, every change on this Pr except last commit that failed to build)

That being said, nothing prevents us to implement the capability for `comment_on_pr --id 'terraform-plan-aws' ""` to delete a comment when the provided comment body is empty (or when we pass `--delete` flag or whatever), and then call that as an in-place replacement of `delete_comment_pr` (ie before running `terraform plan` to ensure old comment got deleted even if `terraform plan` failed. 

It'd just be an implementation change detail (of having this feature from this `delete_pr_comment` being now implemented as `comment_on_pr --delete` instead).
But the big advantage of doing such a change being that we'd now guarantee that the logic to add the `<!-- comment-id:${COMMENT_ID} -->` hidden HTML comment to identify a GH comment would be handled directly by `comment_on_pr` instead of having the caller to do that dance, and then have the guarantee that the logic to find which comment to delete on `comment_on_pr --id xyz --delete` would rely on then same format for the `<!-- comment-id:${COMMENT_ID} -->` bit in both the use case of adding and deleting a comment (instead of having to replicate the HTML comment format assumption in the implementation of both `comment_on_pr --id someid` and `delete_pr_comment --id someid` if we were to keep those feature in separate commands/implementation like here)

Wdyt?

crazytonyli commented:
That sounds good to me! I'll make the change then. It didn't occur to me that we no longer need the `delete_pr_comment`, because `comment_on_pr` with an empty comment does that too.

AliSoftware commented:
Closing the loop: deletion of this new `delete_pr_comment` applied in #53, while the ability to delete a PR comment has been moved into the implementation of #51.
